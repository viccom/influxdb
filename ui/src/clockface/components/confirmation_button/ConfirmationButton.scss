/*
  Confirmation Button Styles
  ------------------------------------------------------------------------------
*/

@import 'src/style/modules';

.confirmation-button {
  position: relative;
}

.confirmation-button--stretch {
  width: 100%;
}

.confirmation-button--tooltip {
  position: absolute;
  top: 100%;
  left: 50%;
  transform: translate(-50%,0);
  transition: all;
  visibility: hidden;
  z-index: 9999;

  &.visible {
    visibility: visible;
  }
}

.confirmation-button--tooltip-body {
  font-size: 13px;
  position: relative;
  padding: $ix-marg-b;
  background-color: $c-curacao;
  border-radius: $radius;
  white-space: nowrap;
  color: $g20-white;
  font-weight: 600;
  user-select: none;
  transition: background-color 0.25s ease;
  box-shadow: 0 0 6px 2px rgba($g0-obsidian, 0.75);
  margin-top: $ix-marg-b;

  &:after {
    content: '';
    position: absolute;
    top: 0;
    left: 50%;
    border: $ix-marg-b solid transparent;
    border-bottom-color: $c-curacao;
    transform: translate(-50%,-100%);
    transition: border-color 0.25s ease;
  }

  &:hover {
    cursor: pointer;
    background-color: $c-dreamsicle;

    &:after {
      border-bottom-color: $c-dreamsicle;
    }
  }
}
