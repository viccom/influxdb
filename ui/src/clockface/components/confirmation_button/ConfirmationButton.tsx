// Libraries
import React, {Component, RefObject} from 'react'
import classnames from 'classnames'

// Components
import Button from 'src/clockface/components/Button/index'
import {ClickOutside} from 'src/shared/components/ClickOutside'

// Types
import {
  ComponentStatus,
  ComponentColor,
  ComponentSize,
  ButtonShape,
  IconFont,
  ButtonType,
} from 'src/clockface/types'

// Styles
import 'src/clockface/components/confirmation_button/ConfirmationButton.scss'

import {ErrorHandling} from 'src/shared/decorators/errors'

interface Props {
  confirmText: string
  onConfirm: (returnValue?: any) => void
  returnValue?: any
  text?: string
  size?: ComponentSize
  shape?: ButtonShape
  icon?: IconFont
  status?: ComponentStatus
  titleText?: string
  tabIndex?: number
  className?: string
}

interface State {
  isTooltipVisible: boolean
}

@ErrorHandling
class ConfirmationButton extends Component<Props, State> {
  public static defaultProps: Partial<Props> = {
    size: ComponentSize.Small,
    shape: ButtonShape.Default,
    status: ComponentStatus.Default,
  }

  public ref: RefObject<HTMLButtonElement> = React.createRef()

  constructor(props: Props) {
    super(props)

    this.state = {
      isTooltipVisible: false,
    }
  }

  public render() {
    const {
      text,
      titleText,
      tabIndex,
      size,
      shape,
      status,
      confirmText,
      icon,
    } = this.props
    const {isTooltipVisible} = this.state

    return (
      <ClickOutside onClickOutside={this.handleHideTooltip}>
        <div className={this.className}>
          <Button
            text={text}
            titleText={titleText}
            tabIndex={tabIndex}
            color={ComponentColor.Danger}
            shape={shape}
            size={size}
            status={status}
            active={isTooltipVisible}
            onClick={this.handleButtonClick}
            icon={icon}
            type={ButtonType.Button}
          />
          <div className={this.tooltipClassName}>
            <div
              data-test="confirmation-button--click-target"
              className="confirmation-button--tooltip-body"
              onClick={this.handleTooltipClick}
            >
              {confirmText}
            </div>
          </div>
        </div>
      </ClickOutside>
    )
  }

  private handleButtonClick = (): void => {
    this.setState({isTooltipVisible: !this.state.isTooltipVisible})
  }

  private handleHideTooltip = (): void => {
    this.setState({isTooltipVisible: false})
  }

  private handleTooltipClick = (): void => {
    const {returnValue, onConfirm} = this.props

    onConfirm(returnValue)
    this.handleHideTooltip()
  }

  private get className(): string {
    const {shape, className} = this.props

    return classnames('confirmation-button', {
      'confirmation-button--stretch': shape === ButtonShape.StretchToFit,
      [className]: className,
    })
  }

  private get tooltipClassName(): string {
    const {isTooltipVisible} = this.state

    return classnames('confirmation-button--tooltip', {
      visible: isTooltipVisible,
    })
  }
}

export default ConfirmationButton
