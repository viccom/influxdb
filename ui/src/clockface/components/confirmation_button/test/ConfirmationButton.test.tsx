import React from 'react'
import {mount} from 'enzyme'

import ConfirmationButton from 'src/clockface/components/confirmation_button/ConfirmationButton'
import Button from 'src/clockface/components/Button'

describe('ConfirmationButton', () => {
  let wrapper

  const wrapperSetup = (override = {}) => {
    const props = {
      text: 'I am a dangerous button!',
      confirmText: 'Click me if you dare',
      onConfirm: () => {},
      ...override,
    }

    return mount(<ConfirmationButton {...props} />)
  }

  it('matches snapshot with minimal props', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('renders correctly', () => {
    wrapper = wrapperSetup()
    expect(wrapper.find(ConfirmationButton)).toHaveLength(1)
  })

  describe('interaction', () => {
    it('shows the tooltip when clicked', () => {
      wrapper = wrapperSetup()

      wrapper.find(Button).simulate('click')

      expect(wrapper).toMatchSnapshot()
    })

    it('returns the specified value on tooltip click', () => {
      const onConfirm = jest.fn(x => x)
      const returnValue = 666

      wrapper = wrapperSetup({returnValue, onConfirm})

      wrapper.find(Button).simulate('click')

      wrapper
        .find({'data-test': 'confirmation-button--click-target'})
        .simulate('click')

      expect(onConfirm.mock.results[0].value).toBe(returnValue)
    })
  })
})
