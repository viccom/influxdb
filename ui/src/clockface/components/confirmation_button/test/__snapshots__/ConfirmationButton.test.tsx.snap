// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationButton interaction shows the tooltip when clicked 1`] = `
<ConfirmationButton
  confirmText="Click me if you dare"
  onConfirm={[Function]}
  shape="none"
  size="sm"
  status="default"
  text="I am a dangerous button!"
>
  <ClickOutside
    onClickOutside={[Function]}
  >
    <div
      className="confirmation-button"
    >
      <Button
        active={true}
        color="danger"
        onClick={[Function]}
        shape="none"
        size="sm"
        status="default"
        text="I am a dangerous button!"
        type="button"
      >
        <button
          className="button button-sm button-danger active"
          disabled={false}
          onClick={[Function]}
          tabIndex={0}
          title="I am a dangerous button!"
          type="button"
        >
          I am a dangerous button!
        </button>
      </Button>
      <div
        className="confirmation-button--tooltip visible"
      >
        <div
          className="confirmation-button--tooltip-body"
          data-test="confirmation-button--click-target"
          onClick={[Function]}
        >
          Click me if you dare
        </div>
      </div>
    </div>
  </ClickOutside>
</ConfirmationButton>
`;

exports[`ConfirmationButton matches snapshot with minimal props 1`] = `undefined`;
